libs/isograph-react/src/core/optimisticProxy.ts

```diff
@@ -0,0 +1,416 @@
+import {
+  callSubscriptions,
+  insertEmptySetIfMissing,
+  type EncounteredIds,
+} from './cache';
+import type {
+  BaseStoreLayerData,
+  IsographEnvironment,
+  StoreLayerData,
+  StoreLink,
+  StoreRecord,
+} from './IsographEnvironment';
+
+export function getOrInsertRecord(
+  dataLayer: StoreLayerData,
+  link: StoreLink,
+): StoreRecord {
+  const recordsById = (dataLayer[link.__typename] ??= {});
+  return (recordsById[link.__link] ??= {});
+}
+
+/**
+ * Given the child-most store layer (i.e. environment.store) and a link (identifying a
+ * store record), create a proxy object that attempts to read through each successive
+ * store layer until a value (i.e. field name) is found. If found, return that value.
+ */
+export function getStoreRecordProxy(
+  storeLayer: StoreLayer,
+  link: StoreLink,
+): StoreRecord {
+  return new Proxy<StoreRecord>(
+    {},
+    {
+      get(_, p) {
+        let node: StoreLayer | null = storeLayer;
+
+        while (node !== null) {
+          const storeRecord = node.data[link.__typename]?.[link.__link];
```

Comment:
Right now, every time we access a field, we re-iterate the layers. Let's keep track of the first storeLayer that contained the store record, as in:

```
let targetStoreLayer: StoreLayer | undefined | null = undefined;

get(_, p) {
  if (targetStoreLayer === undefined) { ... }
  // we still have to do the while loop again
}
set(_,p) {
  targetStoreLayer = storeLayer; // if we set, the record gets added to the childmost layer
}
```